### PR TITLE
Telemetry for Data Workspaces extension

### PR DESCRIPTION
**Pull request by Benjin**
Target telemetry points:
* Existing workspace opened, new workspace created
    * If workspace already opened
    * Number and types of projects in opened workspace
    * Location of .code-workspace file relative to project file(s)
* Projects
    * Added
        * Project/template type
        * Location relative to .code-workspace file
    * Removed
        * Project type
* Registered project providers

### CODE REVIEW

File: extensions/data-workspace/src/common/telemetry.ts [hunk 1]
```diff
@@ -3,15 +3,25 @@
...
 export enum TelemetryViews {
+	WorkspaceTreePane = 'WorkspaceTreePane',
+	OpenDialog = 'OpenDialog',
+	ProviderRegistration = 'ProviderRegistration'
+}
+
+export function CalculateRelativity(projectPath: string, workspacePath?: string): string {
+	vscode.workspace.asRelativePath(projectPath);
+
+	return 'TODO';
```
Comment (Benjin) [inline extensions/data-workspace/src/common/telemetry.ts:26]: Incoming...

File: extensions/data-workspace/src/dataworkspace.d.ts [hunk 2]
```diff
@@ -68,6 +68,11 @@ declare module 'dataworkspace' {
 		 * Gets the supported project types
 		 */
 		readonly supportedProjectTypes: IProjectType[];
+
+		/**
+		 * Gets the extension ID for the project provider
+		 */
+		readonly providerExtensionId: string;
```
Comment (Benjin) [inline extensions/data-workspace/src/dataworkspace.d.ts:75]: Is there a better way to get the full extension name (e.g. "Microsoft.sql-database-projects") without depending on the extension to implement this correctly?

File: extensions/data-workspace/src/common/telemetry.ts [hunk 3]
```diff
@@ -3,15 +3,25 @@
...
-const packageJson = require('../package.json');
+const packageJson = require('../../package.json');
 
 let packageInfo = Utils.getPackageInfo(packageJson)!;
 
 export const TelemetryReporter = new AdsTelemetryReporter(packageInfo.name, packageInfo.version, packageInfo.aiKey);
 
 export enum TelemetryViews {
+	WorkspaceTreePane = 'WorkspaceTreePane',
+	OpenDialog = 'OpenDialog',
```
Comment (kisantia) [inline extensions/data-workspace/src/common/telemetry.ts:19]: OpenExistingDialog

File: extensions/data-workspace/src/dialogs/openExistingDialog.ts [hunk 4]
```diff
@@ -71,10 +77,22 @@ export class OpenExistingDialog extends DialogBase {
...
 			} else {
+				const t = TelemetryReporter.createActionEvent(TelemetryViews.OpenDialog, 'Opening project');
+
+				// save datapoint now because it'll get set to new value during validateWorkspace()
+				t.withAdditionalProperties({ hasWorkspaceOpen: (vscode.workspace.workspaceFile !== undefined).toString() });
+
 				const validateWorkspace = await this.workspaceService.validateWorkspace();
+
 				if (validateWorkspace) {
+					t.withAdditionalProperties({ workspaceProjectRelativity: CalculateRelativity(this._projectFile, this.workspaceInputBox!.value!) }).send();
```
Comment (kisantia) [inline extensions/data-workspace/src/dialogs/openExistingDialog.ts:95]: if they clicked "Cancel" in the validate workspace toast notification confirmation, that should also be reported and sent

File: extensions/data-workspace/src/main.ts [hunk 5]
```diff
@@ -23,18 +24,24 @@ export function activate(context: vscode.ExtensionContext): Promise<IExtension>
 		setProjectProviderContextValue(workspaceService);
 	}));
 	setProjectProviderContextValue(workspaceService);
+
 	context.subscriptions.push(vscode.commands.registerCommand('projects.new', async () => {
+		TelemetryReporter.sendActionEvent(TelemetryViews.WorkspaceTreePane, 'EmptyTreeWelcomeNewButtonClicked');
```
Comment (kisantia) [inline extensions/data-workspace/src/main.ts:29]: doesn't this also get called even when the tree isn't empty?

File: extensions/data-workspace/src/services/workspaceService.ts [hunk 6]
```diff
@@ -115,6 +116,12 @@ export class WorkspaceService implements IWorkspaceService {
 				currentProjects.push(projectFile);
 				newProjectFileAdded = true;
 
+				TelemetryReporter.createActionEvent(TelemetryViews.WorkspaceTreePane, 'add project to workspace')
```
Comment (kisantia) [inline extensions/data-workspace/src/services/workspaceService.ts:119]: why do some action events have spaces in the action name and others don't?

Comment (Benjin) [reply extensions/data-workspace/src/services/workspaceService.ts:119]: Because I don't know what the preferred convention is. 😄

**Comment by Charles-Gagnon** [comment]
@Benjin Looks like your commit history is screwed up. You should do a rebase and reapply of your changes to fix that

File: extensions/data-workspace/src/dataworkspace.d.ts (hunk 2 above)
Comment (Benjin) [reply extensions/data-workspace/src/dataworkspace.d.ts:76]: I'll confirm that this is the case via fiddler, but I think that would always include the `data-workspace` extension ID if it's coming from this extension.  What I'm meaning to capture with this is the set of project-contributing extensions that have registered with `data-workspace`.

File: extensions/data-workspace/package.json [hunk 7]
```diff
@@ -153,7 +153,9 @@
     ]
   },
   "dependencies": {
+    "@opentelemetry/tracing": "^0.14.0",
```
Comment (Charles-Gagnon) [inline extensions/data-workspace/package.json:156]: You should remove these too

File: extensions/data-workspace/src/common/telemetry.ts [hunk 8]
```diff
@@ -3,15 +3,43 @@
...
 export const TelemetryReporter = new AdsTelemetryReporter(packageInfo.name, packageInfo.version, packageInfo.aiKey);
 
 export enum TelemetryViews {
+	WorkspaceTreePane = 'WorkspaceTreePane',
+	OpenExistingDialog = 'OpenExistingDialog',
+	NewProjectDialog = 'NewProjectDialog',
+	ProviderRegistration = 'ProviderRegistration'
+}
+
+export function CalculateRelativity(projectPath: string, workspacePath?: string): string {
```
Comment (Charles-Gagnon) [inline extensions/data-workspace/src/common/telemetry.ts:25]: nit: Function names should be camelCased

File: extensions/data-workspace/src/dataworkspace.d.ts (hunk 2 above)
Comment (Benjin) [reply extensions/data-workspace/src/dataworkspace.d.ts:76]: Confirmed: the extension ID that's automatically captured isn't the one I'm intending to here.

**Review by kisantia — APPROVED** [review]

**Pull request merged**